File: Backtracking_and_Search/LC_46_Permutations.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        used = [False] * n
        ans = []
        def dfs(curr, ans):
            if len(curr) == n:
                ans.append(curr[:])
                return 
            for i in range(n):
                if used[i]: continue
                used[i] = True
                curr.append(nums[i])
                dfs(curr, ans)
                used[i] = False
                curr.pop()
        dfs([], ans)
        return ans

File: Backtracking_and_Search/LC_46_Permutations.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        vector<int> used(n, 0);
        vector<int> curr;
        dfs(nums, curr, used, ans);
        return ans;
    }
private:
    void dfs(vector<int>& nums, vector<int>& curr, vector<int>& used, vector<vector<int>>& ans){
        if(curr.size() == nums.size()){
            ans.push_back(curr);
            return;
        }
        for (int i = 0; i < nums.size(); i++){
            if (used[i]){
                continue;
            }
            used[i] = 1;
            curr.push_back(nums[i]);
            dfs(nums, curr, used, ans);
            curr.pop_back();
            used[i] = 0;
        }
    }
};